kf/nfconfig.go

```diff
@@ -942,3 +943,367 @@ func getHostInterfaces() (map[string]bool, error) {
+	}
+	if err := c.SaveConfigsToConfigStore(); err != nil {
+		return fmt.Errorf("AddeBPFPrograms failed to save configs %v", err)
+	}
+	return nil
+}
+
+// DeleteProgramsOnInterface : It will delete ebpf Programs on the given interface
+func (c *NFConfigs) DeleteProgramsOnInterface(ifaceName, HostName string, bpfProgs *models.BPFProgramNames) error {
+	if HostName != c.HostName {
+		errOut := fmt.Errorf("provided bpf programs do not belong to this host")
+		log.Error().Err(errOut)
+		return errOut
+	}
+
+	if ifaceName == "" || bpfProgs == nil {
+		errOut := fmt.Errorf("iface name or bpf programs are empty")
+		log.Error().Err(errOut)
+		return errOut
+	}
+
+	if _, ok := c.hostInterfaces[ifaceName]; !ok {
+		errOut := fmt.Errorf("%s interface name not found in the host", ifaceName)
+		log.Error().Err(errOut)
+		return errOut
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	sort.Strings(bpfProgs.XDPIngress)
+	if c.IngressXDPBpfs[ifaceName] != nil {
+		bpfList := c.IngressXDPBpfs[ifaceName]
+		for e := bpfList.Front(); e != nil; {
+			next := e.Next()
+			data := e.Value.(*BPF)
+			if BinarySearch(bpfProgs.XDPIngress, data.Program.Name) {
+				err := c.DeleteProgramsOnInterfaceHelper(e, ifaceName, models.XDPIngressType, bpfList)
+				if err != nil {
+					return fmt.Errorf("DeleteProgramsOnInterfaceHelper function failed : %v", err)
+				}
+			}
+			e = next
+		}
+	}
+
+	sort.Strings(bpfProgs.TCIngress)
+	if c.IngressTCBpfs[ifaceName] != nil {
+		bpfList := c.IngressTCBpfs[ifaceName]
+		for e := bpfList.Front(); e != nil; {
+			next := e.Next()
+			data := e.Value.(*BPF)
+			if BinarySearch(bpfProgs.TCIngress, data.Program.Name) {
+				err := c.DeleteProgramsOnInterfaceHelper(e, ifaceName, models.IngressType, bpfList)
+				if err != nil {
+					return fmt.Errorf("DeleteProgramsOnInterfaceHelper function failed : %v", err)
+				}
+			}
+			e = next
+		}
+	}
+
+	sort.Strings(bpfProgs.TCEgress)
+	if c.EgressTCBpfs[ifaceName] != nil {
+		bpfList := c.EgressTCBpfs[ifaceName]
+		for e := bpfList.Front(); e != nil; e = e.Next() {
+			next := e.Next()
+			data := e.Value.(*BPF)
+			if BinarySearch(bpfProgs.TCEgress, data.Program.Name) {
+				err := c.DeleteProgramsOnInterfaceHelper(e, ifaceName, models.EgressType, bpfList)
+				if err != nil {
+					return fmt.Errorf("DeleteProgramsOnInterfaceHelper function failed : %v", err)
+				}
+			}
+			e = next
+		}
+	}
+	return nil
+}
+
+// DeleteProgramsOnInterfaceHelper : helper function for DeleteProgramsOnInterface function
+func (c *NFConfigs) DeleteProgramsOnInterfaceHelper(e *list.Element, ifaceName string, direction string, bpfList *list.List) error {
+	if e == nil {
+		return nil
+	}
+	prog := e.Value.(*BPF)
+	prog.Program.AdminStatus = models.Disabled
+	if err := prog.Stop(ifaceName, direction, c.HostConfig.BpfChainingEnabled); err != nil {
+		return fmt.Errorf("failed to stop %s iface %s direction %s", prog.Program.Name, ifaceName, direction)
+	}
+	tmpNextBPF := e.Next()
+	tmpPreviousBPF := e.Prev()
```

Comment:
the tmpNextBPF and tmpPreviousBPF lines seem to make more sense to put after line 1258, as this doesn't make sense if BpfChainingEnabled is disabled.

Reply:
It will break  because we are removing **e from bpfList**